send-emails-from-a-service/main.bal

```diff
@@ -0,0 +1,172 @@
+
+    resource function post categories/[string category]/reserve(ReservationRequest payload)
+            returns http:InternalServerError|http:NotFound|http:Ok {
+
+        ReservationRequest {
+            patient: {cardNo, ...patient},
+            doctor,
+            hospital,
+            hospital_id,
+            appointment_date
+        } = payload;
+
+        Appointment|http:ClientError appointment =
+                hospitalServicesEP->/[hospital_id]/categories/[category]/reserve.post({
+            patient,
+            doctor,
+            hospital,
+            appointment_date
+        });
+
+        if appointment !is Appointment {
+            log:printError("Appointment reservation failed", appointment);
+            if appointment is http:ClientRequestError {
+                return <http:NotFound> {body: string `unknown hospital, doctor, or category`};
+            }
+            return <http:InternalServerError> {body: appointment.message()};
+        }
+
+        int appointmentNumber = appointment.appointmentNumber;
+
+        Payment|http:ClientError payment = paymentEP->/.post({
+            appointmentNumber: appointmentNumber,
+            doctor: appointment.doctor,
+            patient: appointment.patient,
+            fee: appointment.doctor.fee,
+            confirmed: appointment.confirmed,
+            card_number: cardNo
+        });
+
+        if payment !is Payment {
+            log:printError("Payment settlement failed", payment);
+            if payment is http:ClientRequestError {
+                return <http:NotFound> {body: string `unknown appointment number`};
+            }
+            return <http:InternalServerError> {body: payment.message()};
+        }
+
+        string messageContent = string `Appointment Confirmation
+
+        Appointment Details
+            Appointment Number : ${appointmentNumber}
+            Appointment Date: ${appointment_date}
+
+        Patient Details
+            Name : ${patient.name}
+            Contact Number : ${patient.phone}
+
+        Doctor Details
+            Name : ${appointment.doctor.name}
+            Specialization : ${appointment.doctor.category}
+
+        Payment Details
+            Doctor Fee : ${payment.actualFee}
+            Discount : ${payment.discount}
+            Total Fee : ${payment.discounted}
+            Payment Status : ${payment.status}`;
+
+        email:Error? sendMessage = smtpClient->sendMessage({
+            to: patient.email,
+            subject: "Appointment reservation confirmed at " + hospital,
+            body: messageContent
+        });
+        
+        if sendMessage is email:Error {
+            return <http:InternalServerError> {body: sendMessage.message()};
+        }
+        log:printDebug("Email sent successfully",
+                       name = patient.name,
+                       appointmentNumber = appointmentNumber);
+        return <http:Ok> {body: messageContent};
```

Comment:
Best practice for POST maybe 201. https://ballerina.io/spec/http/#2353-default-response-status-codes 

May be able to omit the body too.

Reply:
Since I'm testing the message content, i won't be able to omit the body

Reply:
We shouldn't introduce actual logic just for testing purposes. (I know we do something similar for client initialization, but that isn't quite the same and behaviour doesn't change).

You can test the body during the mocking of the send email action and/or a separate unit test once https://github.com/ballerina-guides/integration-tutorials/pull/7#discussion_r1328397262 is done.